process/estimator/executionResultInclusionEstimator.go

```diff
@@ -0,0 +1,189 @@
+package estimator
+
+import (
+	"math/bits"
+
+	logger "github.com/multiversx/mx-chain-logger-go"
+)
+
+var log = logger.GetOrCreate("process/executionResultInclusionEstimator")
+
+// ExecutionResultMeta is a lightweight summary EIE requires.
+type ExecutionResultMeta struct {
+	HeaderHash   [32]byte // Link to full header in DB / cache
+	HeaderNonce  uint64   // Monotonic within shard
+	HeaderTimeMs uint64   // Milliseconds since Unix epoch, from header
+	GasUsed      uint64   // Units actually consumed (post‑execution)
+}
+
+// Config supplied at construction, read‑only thereafter.
+// TODO add also max estimated block gas capacity  // used gas must be lower than this
+type Config struct {
+	SafetyMargin       uint64 // default 110
+	MaxResultsPerBlock uint64 // 0 = unlimited
+	GenesisTimeMs      uint64 // required if lastNotarised == nil
+}
+
+// ExecutionResultInclusionEstimator (EIE) is a deterministic component shipped with the MultiversX *Supernova*
+// node. It determines, at proposal‑time and at validation‑time, whether one or more pending execution results can be
+// safely embedded in the block that is being produced / verified.
+type ExecutionResultInclusionEstimator struct {
+	cfg  Config // immutable after construction
+	tGas uint64 // time per gas unit on **minimum‑spec** hardware - 1 ns per gas unit
+	// TODO add also max estimated block gas capacity  // used gas must be lower than this
+}
+
+// NewExecutionResultInclusionEstimator returns a new instance of EIE
+func NewExecutionResultInclusionEstimator(cfg Config) *ExecutionResultInclusionEstimator {
+	return &ExecutionResultInclusionEstimator{
+		cfg:  cfg,
+		tGas: uint64(1),
+	}
+}
+
+// SetTimePerGasUnit sets the time per gas unit on minimum‑spec hardware.
+func (erie *ExecutionResultInclusionEstimator) SetTimePerGasUnit(tGas uint64) {
+	if tGas == 0 {
+		log.Warn("ExecutionResultInclusionEstimator: SetTimePerGasUnit called with zero value, using default 1 ns per gas unit")
+		tGas = 1
+	}
+	if erie.tGas != tGas {
+		log.Debug("ExecutionResultInclusionEstimator: SetTimePerGasUnit called",
+			"oldValue", erie.tGas,
+			"newValue", tGas)
+		erie.tGas = tGas
+	}
+}
+
+// Decide returns the prefix of `pending` that may be inserted into the block currently being built / verified.
+// Return value: `allowed` is the count of leading entries in `pending` deemed safe. The caller slices `pending[:allowed]`and embeds them.
+func (erie *ExecutionResultInclusionEstimator) Decide(lastNotarised *ExecutionResultMeta,
+	pending []ExecutionResultMeta,
+	currentHdrTsNs uint64,
```

Comment:
I think it would be better to have this as ms. The header timestamp will be in milli seconds, and I think it would be better that it is simply passed and all further conversions needed are done inside the estimation component.
(we need to update also the specs)